### DISCUSSION TITLE
Embeddings / feature vectors

### DISCUSSION THREAD

**Discussion by arjenpdevries** [duckdb/duckdb#3811] [General]
# Embeddings / feature vectors

We are looking into using DuckDB in various NLP and IR applications. Modern NLP makes extensive use of representations of text in high-dimensional real spaces, using what they call embeddings.

From our DB perspective, these are just long rows of numbers (usually doubles / floats / decimals). Two important operations are to get a token's or word's embedding to use in further processing (this can be a simple point lookup, but often we need all embeddings for a short piece of text, so a rather selective join) and to perform similarity search (find embeddings near a given one).

In theory, using DuckDB as an IR backend should let us keep all the work with both text and embeddings in a single data space. However I run into various problems. In this discussion, I hope to get some guidance how best to proceed.

Our first idea was to simply create a table with the N columns for the dimensionality of the embeddings (in the order of 200-300). This does not work very well - this makes sense, because DuckDB has to re-combine data from many different columns (column segments) to reconstruct the feature vector (embedding) we want to use in further processing.

The second idea was to use the `STRUCT` data type to add a row-oriented representation to the column. We could then still run algorithms that need only a few columns (like Bond) but opt to use the column with the structs to get quick access to complete embeddings (for further processing, in Python / PyTorch to be precise).

Let's take a look at what happens:

```
import numpy
import pandas
import duckdb

con = duckdb.connect()
con.execute("CREATE TABLE emb(d1 DOUBLE, d2 DOUBLE, d3 DOUBLE)")
con.execute("INSERT INTO emb VALUES (1.0, 0.8, 0.3)")
con.execute("INSERT INTO emb VALUES (0.9, 0.9, 0.1)")
con.execute("INSERT INTO emb VALUES (0.0, 0.7, 0.2)")

results = con.execute("SELECT (d1,d2,d3) FROM emb;").fetchnumpy()
print("Numpy results:")
print(results)

results = con.execute("SELECT (d1,d2,d3) FROM emb;").fetchdf()
print("Pandas results:")
print(results)
```

The output is very nice for some applications...
```
Numpy results:
{'main.row(d1, d2, d3)': array([{'d1': 1.0, 'd2': 0.8, 'd3': 0.3},
       {'d1': 0.9, 'd2': 0.9, 'd3': 0.1},
       {'d1': 0.0, 'd2': 0.7, 'd3': 0.2}], dtype=object)}
Pandas results:
                main.row(d1, d2, d3)
0  {'d1': 1.0, 'd2': 0.8, 'd3': 0.3}
1  {'d1': 0.9, 'd2': 0.9, 'd3': 0.1}
2  {'d1': 0.0, 'd2': 0.7, 'd3': 0.2}
```

... but not for ours!

I suspect that the internal representation in DuckDB would be what we need: one column, with a metadata row with the keys, and many rows with just the three numbers each, stored consecutively in memory. Unfortunately, this internal representation - if I am making the right assumptions - is converted into python objects at the deepest level, where we would just want to obtain an array of arrays of floats instead.

Are we abusing the STRUCT datatype? Or are my assumptions correct and is it only the interfacing to the Python module that would need to be modified? What would be a good API to determine how to represent the DuckDB STRUCT in Python - I expect that my preferred representation is not good for many other applications. Do I overlook an option to achieve the same already, perhaps through using PyArrow?

_(Bottom line, we fall back to our initial use of Sqlite in this case - using it is at least 10x faster, in spite of obvious overheads for squeezing our embeddings into it. It would however be much preferred to achieve our objective with DuckDB of course! I also worry that we simply use DuckDB wrong...)_

**Comment by lnkuiper** [comment]
Internally, each struct field is just another column, i.e., a struct vector just has a few child vectors (3 in your case). This makes packing / extracting structs essentially free, because we don't need to change the data at all to extract a field from a struct, we can simply reference it. In short, selecting the struct `(d1, d2, d3)` does not change the in-memory representation much, it's still columnar.

If you need the values `d1, d2, d3` to be stored consecutively, in memory, your only option is to use the `LIST` type. The `LIST` type is also columnar, but the values in the lists that belong to consecutive rows are stored consecutively in memory. 

I would suggest using the following query:
```python
results = con.execute("SELECT [d1,d2,d3] FROM emb;")
```
 
Which will result in the following in-memory representation in DuckDB:
```
[ row1.d1, row1.d2, row1.d3, row2.d1, row2.d2, ..., row3.d3 ]
```

If you call `fetchdf()`/`fetchnumpy()`, you get Python lists:
```
  main.list_value(d1, d2, d3)
0             [1.0, 0.8, 0.3]
1             [0.9, 0.9, 0.1]
2             [0.0, 0.7, 0.2]
```

**Reply by arjenpdevries** [reply]
So we tried that already before, but creating the list representation caused O-O-M errors. Also, I guess we would better off to create a fixed-size column format that is simpler than the variable-length LIST representation? (I thought I read that LIST uses a C++ Vector for every list? Maybe misunderstood?)

**Reply by Mytherin** [reply]
The `LIST` aggregate function uses a Vector for every entry in the hash table, which means that doing a query like `SELECT grp, LIST(val) FROM tbl GROUP BY grp` can be expensive if there are many unique groups. The `LIST` type internally groups together lists in a columnar format and should be rather fast.

**Reply by lnkuiper** [reply]
The `list` aggregate function indeed uses a lot of memory, and we are looking at ways to improve that #3734.
However, the `[]` notation uses the `list_value` function, which is just a scalar function. If that creates OOM errors I'd be surprised, because it should use about the same memory as the struct representation.

**Comment by Alex-Monahan** [comment]
It is not directly related to your main question, but have you tried using our full text search extension to help with your similarity search?
https://duckdb.org/docs/extensions/full_text_search

**Reply by lnkuiper** [reply]
Thanks for promoting my extension, Alex! @arjenpdevries actually wrote the [paper that we implemented in our FTS extension](https://dl.acm.org/doi/abs/10.1145/2600428.2609460) together with Hannes, just so you know :)

**Reply by Alex-Monahan** [reply]
Ah, sorry about that! Thank you for the background!

**Reply by arjenpdevries** [reply]
Yeah, thanks for the tip :-) For similarity search over the text, that is a cool approach, but k-NN similarity search in embedding space is a different problem and the extension does not help. I want to revive the [BOND paper](https://dblp.org/rec/conf/sigmod/VriesMNK02.html) that we once did for this purpose, but that is for later this year.

**Comment by michalwols** [comment]
Related to this I'd love to see support for approximate nearest neighbor indexing

**Reply by changhiskhan** [reply]
Hey Michal! We're going to be working on this in Lance after data versioning. We've played with implementing distance functions and polygon contains in a duckdb extension if you're interested in kicking the tires.

**Comment by vimota** [comment]
Would love to hear if someone has gotten a use case like this to work, I've used Pinecone in the past for Vector/embedding similarity search, but would love to have that in DuckDB so we could further filter on other metadata.

**Comment by lnkuiper** [comment]
No update on this yet. We have thought about it though, and DuckDB's internal list representation is convenient and efficient - this is not the problem. The problem is that our lists can be variable-size, and therefore when we convert the query result to Python, we cannot use a numpy array, and are forced to use Python's lists instead. This is, of course, much slower.

If we can somehow determine that all the lists are the same size, we can convert to a 2D numpy array. This would probably require a new `LogicalType`: "fixed-size lists", and some checks within DuckDB to make sure that this fixed-size is enforced.

**Reply by changhiskhan** [reply]
Yeah, echo the need for fixed sized lists in duckdb. Arrow already has fixed sized lists - not sure if that helps at all with testing the performance in duckdb. Would it make sense to do some experiments to create a "thin type" in duckdb that basically just push list operations down Arrow compute kernels?

In addition to fixed sized lists, we need to have new index types that does vector quantization and then provides indexing into the table itself. We're working on this in Lance OSS (https://github.com/eto-ai/lance) and it'll be transparently accessible via Arrow scanner API if your pyarrow table is serialized as Lance format.

Performance-wise we've played with duckdb extensions that implement vector distance functions. For reasonable sized data duckdb already does pretty decently at this. Others have experimented with doing pre-hashing and for that you don't even need to write C++ code (trade off is you need pre-processing).

**Reply by changhiskhan** [reply]
Oh also, if you're able to support N-dimensional fixed sized lists then we can dress up images, tensors, and videos clips as duckdb types without copying ;)  I've got a replacement scan hack for mp4 files and I'm passing around frames as structs with raw bytes and dimensions 🤦🏻‍♂️

**Reply by sean-lynch** [reply]
Purely for folks following along on this discussion, here's a related conversation with the pandas team about moving off numpy to arrow to handle some of these performance issues: https://datapythonista.me/blog/pandas-20-and-the-arrow-revolution-part-i

**Reply by Tishj** [reply]
When I saw new activity here I was expecting it to be about one of the latest PRs that significantly reduces the size overhead of lists

**Comment by changhiskhan** [comment]
> # Embeddings / feature vectors
> We are looking into using DuckDB in various NLP and IR applications. Modern NLP makes extensive use of representations of text in high-dimensional real spaces, using what they call embeddings.
> 
> From our DB perspective, these are just long rows of numbers (usually doubles / floats / decimals). Two important operations are to get a token's or word's embedding to use in further processing (this can be a simple point lookup, but often we need all embeddings for a short piece of text, so a rather selective join) and to perform similarity search (find embeddings near a given one).
> 
> In theory, using DuckDB as an IR backend should let us keep all the work with both text and embeddings in a single data space. However I run into various problems. In this discussion, I hope to get some guidance how best to proceed.
> 
> Our first idea was to simply create a table with the N columns for the dimensionality of the embeddings (in the order of 200-300). This does not work very well - this makes sense, because DuckDB has to re-combine data from many different columns (column segments) to reconstruct the feature vector (embedding) we want to use in further processing.
> 
> The second idea was to use the `STRUCT` data type to add a row-oriented representation to the column. We could then still run algorithms that need only a few columns (like Bond) but opt to use the column with the structs to get quick access to complete embeddings (for further processing, in Python / PyTorch to be precise).
> 
> Let's take a look at what happens:
> 
> ```
> import numpy
> import pandas
> import duckdb
> 
> con = duckdb.connect()
> con.execute("CREATE TABLE emb(d1 DOUBLE, d2 DOUBLE, d3 DOUBLE)")
> con.execute("INSERT INTO emb VALUES (1.0, 0.8, 0.3)")
> con.execute("INSERT INTO emb VALUES (0.9, 0.9, 0.1)")
> con.execute("INSERT INTO emb VALUES (0.0, 0.7, 0.2)")
> 
> results = con.execute("SELECT (d1,d2,d3) FROM emb;").fetchnumpy()
> print("Numpy results:")
> print(results)
> 
> results = con.execute("SELECT (d1,d2,d3) FROM emb;").fetchdf()
> print("Pandas results:")
> print(results)
> ```
> 
> The output is very nice for some applications...
> 
> ```
> Numpy results:
> {'main.row(d1, d2, d3)': array([{'d1': 1.0, 'd2': 0.8, 'd3': 0.3},
>        {'d1': 0.9, 'd2': 0.9, 'd3': 0.1},
>        {'d1': 0.0, 'd2': 0.7, 'd3': 0.2}], dtype=object)}
> Pandas results:
>                 main.row(d1, d2, d3)
> 0  {'d1': 1.0, 'd2': 0.8, 'd3': 0.3}
> 1  {'d1': 0.9, 'd2': 0.9, 'd3': 0.1}
> 2  {'d1': 0.0, 'd2': 0.7, 'd3': 0.2}
> ```
> 
> ... but not for ours!
> 
> I suspect that the internal representation in DuckDB would be what we need: one column, with a metadata row with the keys, and many rows with just the three numbers each, stored consecutively in memory. Unfortunately, this internal representation - if I am making the right assumptions - is converted into python objects at the deepest level, where we would just want to obtain an array of arrays of floats instead.
> 
> Are we abusing the STRUCT datatype? Or are my assumptions correct and is it only the interfacing to the Python module that would need to be modified? What would be a good API to determine how to represent the DuckDB STRUCT in Python - I expect that my preferred representation is not good for many other applications. Do I overlook an option to achieve the same already, perhaps through using PyArrow?
> 
> _(Bottom line, we fall back to our initial use of Sqlite in this case - using it is at least 10x faster, in spite of obvious overheads for squeezing our embeddings into it. It would however be much preferred to achieve our objective with DuckDB of course! I also worry that we simply use DuckDB wrong...)_

@arjenpdevries Leo Hentschker (Column) had tweeted some experiments he did here: https://colab.research.google.com/drive/16uWjlW7z2sYUDNW3yiDPRLG4kMjU-gcH#scrollTo=3ixSa0mUnJYd

good performance but requires some hashing tricks

**Comment by psychemedia** [comment]
In passing, I note a workaround for vector similarity searches around embeddings stored in postgres tables using the duckdb postgres scanner here: https://blog.eto.ai/vector-similarity-search-with-duckdb-44dec043532a

**Comment by wenerme** [comment]
Maybe just integrate https://github.com/facebookresearch/faiss some how ?

**Reply by ned2** [reply]
[hnswlib](https://github.com/nmslib/hnswlib) could be worth considering. It's a lightweight C++ header-only HNSW implementation, one of the more popular ANN algorithms, and has no dependencies, so has been used by a few vector DBs
